docs/modules/ROOT/pages/advanced-scripting.adoc

```diff
@@ -0,0 +1,749 @@
+print("  ", took//microsecond, "microseconds")
+print("  ", took//millisecond, "milliseconds")
+print("  ", took//second, "seconds")
+----
+
+Running it outputs:
+----
+The operation took:
+   4152911 nanoseconds
+   4152 microseconds
+   4 milliseconds
+   0 seconds
+----
+
+==== sleep
+
+It may sometimes be necessary to wait a bit before doing other work.
+`sleep` command waits for the given time in nanoseconds.
+It's very easy to use other time units, such as milliseconds and seconds.
+See the link:clc-script.adoc[script] command documentation for all time units.
+Here's a script that just waits for 2 seconds.
+
+[source,python]
+----
+print("Falling asleep...")
+sleep(2*second)
+print("Woke up")
+----
+
+
+
+=== Logging
+
+CLC provides several logging functions, which log messages using CLC's logging mechanism.
+
+* `log_error(STRING_OR_ERROR)`: Logs a message or an error with the ERROR level.
+* `log_warn(STRING)`: Logs a message with the WARN level.
+* `log_info(STRING)`: Logs a message with the INFO level.
+* `log_debug(STRING)`: Logs a message with the DEBUG level.
+
+If the logging level set with `--log.level` is above the message's log level, then it is not written to the log.
+For example, if `--log.level error` specified then only `log_error` functions log messages.
+
+The `--log.path PATH` flag works as expected and can be used to redirect the log messages to the given file or to STDERR (usually the terminal) if `stderr` is used.
+
+Here is a sample script to try the interplay of the log messages and logging levels.
+Save it as `logging.star`.
+
+[source,python]
+----
+log_error("This is an error log")
+log_warn("This is is a warn log")
+log_info("This is an info log")
+log_debug("This is a debug log")
+----
+
+First, run it as follows.
+Note that we set `--log.path stderr` to see the logs on the screen.
+Otherwise they would written in the default log file.
+
+[source, bash]
+----
+clc script --log.path stderr --log.level error logging.star
+----
+
+Since only log messages above ERROR are to be displayed, you should get output similar to the following:
+
+----
+2023-11-01T11:24:18+03:00  ERROR  [logging.star:2:10] This is an error log
+----
+
+Let's try once more with a different log level.
+This time with DEBUG:
+
+[source, bash]
+----
+clc script --log.path stderr --log.level debug logging.star
+----
+
+This time all log messages should be displayed and you should get output similar to the following:
+
+----
+2023-11-01T21:30:13+03:00  ERROR  [logging.star:2:10] This is an error log
+2023-11-01T21:30:13+03:00   WARN  [logging.star:3:9] This is is a warn log
+2023-11-01T21:30:13+03:00   INFO  [logging.star:4:9] This is an info log
+2023-11-01T21:30:13+03:00  DEBUG  [logging.star:5:10] This is a debug log
+----
+
+=== Debugging
+
+Debug functions start with double underscores (`__`).
+They run only when the debug mode is activated by passing `--debug` flag to the `script` command.
+If the debug mode is not active, they are no-ops.
```

Comment:
"they are no-ops"? what are? The debug functions?

Are you just re-stating that debug mode must be active to use debug functions?

Reply:
I meant, if the debug mode is not active, the debug functions are equivalent to functions doing nothing. The user still can have them in their script though.

Reply:
I'd suggest changing "no-ops" to "non-operational; however, you can keep them in your script."

If no-ops is a commonly-understood term, it can stay as-is; however, my searches only gave me no-ops as an alternative to dev-ops, which is a different thing entirely

Reply:
I used "no-ops" as the plural of "no-op" which stands for "no operation":
* https://english.stackexchange.com/questions/25993/what-does-no-op-mean
* https://en.wikipedia.org/wiki/NOP_(code)

Does that make sense? I will change it if it doesn't look good or may not be understood well.

Reply:
OK, how about providing the Wiki link so that people can find out if they don't know? Along the lines of:

If debug mode is not active, debug functions are https://en.wikipedia.org/wiki/NOP_(code)[no-op^].

Reply:
I think that would confuse our audience more :)
I'll just change the sentence to:

> They are ignored if the debug mode is not active.

if that sounds OK.

Reply:
Thanks, updated at: https://github.com/hazelcast/clc-docs/pull/4/commits/913827092bc66487298846a4601f01becab9796d